source/comms/src/controller.rs

```diff
@@ -324,6 +342,59 @@ async fn complete_pendings<T: FrameSerial>(
     Ok(())
 }
 
+/// A helper function for moving targets from the Known stage to either the Active stage or to the Free stage
+async fn update_known<T: FrameSerial>(
+    inner: &mut [Peer; MAX_TARGETS],
+    serial: &mut T,
+) -> Result<(), Error<T::SerError>> {
+    for (i, p) in inner.iter_mut().enumerate() {
+        // Only worry about pending nodes
+        let Some((mac, instant)) = p.is_known() else {
+            continue;
+        };
+
+        if instant + KNOWN_TIMEOUT >= Instant::now() {
+            p.reset_to_free();
+            continue;
+        }
+
+        // Send a message with the expected MAC address for confirmation
+        let mut out_buf = [0u8; 9];
+        out_buf[0] = CmdAddr::DiscoverySuccess(i as u8).into();
+        out_buf[1..9].copy_from_slice(&mac.to_le_bytes());
+
+        // We should only get back an empty ACK and nothing else
+        let mut in_buf = [0u8; 2];
+        serial.send_frame(&out_buf).await?;
+        let rxto = with_timeout(REPLY_TIMEOUT, serial.recv(&mut in_buf));
+
+        match rxto.await {
+            Ok(Ok(tf)) => {
+                let frame = tf.frame;
+                let good_len = frame.len() == 1;
+                let good_hdr = good_len && frame[0] == CmdAddr::ReplyFromAddr(i as u8).into();
+                if good_hdr {
+                    nut_info!("Promoting to active {=usize} {=u64}", i, mac);
+                    p.promote_to_active();
+                } else {
+                    p.increment_error();
```

Comment:
We could actually cancel the known state earlier on error, but since it has a timeout anyway I'm tempted to just ignore them